addons/stock_picking_batch/models/stock_move_line.py

```diff
@@ -89,3 +91,258 @@ def _add_to_wave(self, wave=False):
         if picking_to_wave_vals_list:
             self.env['stock.picking'].create(picking_to_wave_vals_list)
         wave.action_confirm()
+
+    def _is_auto_waveable(self):
+        self.ensure_one()
+        if not self.picking_id \
+           or self.picking_id.state != 'assigned' \
+           or float_is_zero(self.quantity, precision_rounding=self.product_uom_id.rounding) \
+           or self.batch_id.is_wave \
+           or not self.picking_type_id._is_auto_wave_grouped() \
+           or (self.picking_type_id.wave_group_by_category and self.product_id.categ_id not in self.picking_type_id.wave_category_ids):  # noqa: SIM103
+            return False
+        return True
+
+    def _auto_wave(self):
+        """ Try to find compatible waves to attach the move lines to, otherwise create new waves when possible/appropriate. """
+        lines_nearest_parent_locations = defaultdict(lambda: self.env['stock.location'])
+        batchable_line_ids = OrderedSet()
+        for line in self:
+            if not line._is_auto_waveable():
+                continue
+            if not line.picking_type_id.wave_group_by_location:
+                batchable_line_ids.add(line.id)
+                continue
+            # We want to find the most descendant location in the wave locations list that is a parent of the line location.
+            # Since the wave locations are ordered by complete_name (from the most descendant to the most ancestor), we can iterate in reverse order.
+            for location in reversed(line.picking_type_id.wave_location_ids):
```

Comment:
It will be an issue with a lot of lines and locations. You will probably need to build a tree of locations or something like that to avoid the O(n²)

Reply:
But we are only iterating on the locations defined by the user in the `many2many` field not all the locations, so couldn't we assume that the number of locations is of constant complexity?

Reply:
Yeah but we have the number of lines * number of locations in the m2m * string matching complexity
I would go on
```python
wave_loc_by_picking_type = {}
for picking_type in self.picking_type_id:
    if not picking_type_id.wave_group_by_location:
        continue
    if picking_type in wave_loc_by_picking_type:
        continue
    wave_loc_by_picking_type[picking_type] = set(picking_type.wave_location_ids.ids)

for line in self:
    ...
    wave_loc_set = wave_loc_by_picking_type[line.picking_type_id]
    loc = line.location_id
    while(loc):
        if loc.id in wave_loc_by_picking_type:
            lines_nearest_parent_locations[line] = loc
            break
        loc = loc.parent_id
```
It won't reduce the global complexity but it should provide a better average.
We won't browse all the location each time but just check if the location (or it's parent) is in the m2m.